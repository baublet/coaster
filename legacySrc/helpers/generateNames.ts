import { camelCase, pascalCase, snakeCase } from "change-case";
import pluralize from "pluralize";

export interface GeneratedNames {
  /**
   * The canonical name. Not safe for use in databases or in JS.
   */
  canonical: string;
  original: string;
  /**
   * Pluralized version of original.
   */
  originalPlural: string;
  /**
   * A database- and javascript-safe version of the string. Suitable for using
   * as database column names.
   */
  safe: string;
  /**
   * The plural version of the safe name.
   */
  safePlural: string;
  pascal: string;
  pascalPlural: string;
  camel: string;
  camelPlural: string;
  snake: string;
  snakePlural: string;
}

/**
 * Takes a string and returns a structure of various names for use within an
 * application. If `original` is passed in, we use that as the original value.
 * This is useful if the name passed in is already transformed in some way.
 * @param name
 * @param original
 */
export function generateNames(name: string): GeneratedNames {
  const snake = snakeCase(name);
  const pascal = pascalCase(name);
  const camel = camelCase(name);
  const plural = pluralize(name);
  return {
    canonical: pascal,
    original: name,
    originalPlural: plural,
    safe: snake,
    safePlural: snakeCase(plural),
    pascal: pascal,
    pascalPlural: pascalCase(plural),
    camel: camel,
    camelPlural: camelCase(plural),
    snake,
    snakePlural: snakeCase(plural),
  };
}
