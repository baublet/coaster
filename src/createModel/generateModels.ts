import knex from "knex";
import DataLoader from "dataloader";

type ModelFactoryFactory<TModel = {}, TRawModel = {}, TConnection = any> = (
  getConnection: () => Promise<TConnection>
) => ModelFactory<TModel, TRawModel, TConnection>;

interface ModelFactory<
  TModel = {},
  TRawModel = {},
  TConnection = any,
  TQuery = any
> {
  getConnection: () => Promise<TConnection>;
  getDataLoader: <
    TKey extends keyof TRawModel,
    TMany extends boolean | undefined
  >(
    key: keyof TRawModel,
    many?: TMany
  ) => Promise<
    DataLoader<TRawModel[TKey], TMany extends true ? TRawModel[] : TRawModel>
  >;
  getQuery: () => Promise<TQuery>;
  create: (args: Partial<TRawModel>) => Promise<TRawModel>;
  createMany: (args: Partial<TRawModel>[]) => Promise<TRawModel[]>;
  update: (args: Partial<TRawModel>) => Promise<TRawModel>;
  updateMany: (args: Partial<TRawModel>) => Promise<TRawModel>;
  upsert: (args: Partial<TRawModel>) => Promise<TRawModel>;
  upsertMany: (args: Partial<TRawModel[]>) => Promise<TRawModel[]>;
}

interface Schema {
  name: string;
  models: SchemaModel[];
  modelFactory: ModelFactoryFactory;
}

interface SchemaModel {
  id: string;
  names: {
    singularCamel: string;
    singularPascal: string;
    pluralCamel: string;
    pluralPascal: string;
  };
  properties: Record<string, ModelPropertyDefinition>;
}

interface ModelDefinition {
  names: {
    singularCamel: string;
    singularPascal: string;
    pluralCamel: string;
    pluralPascal: string;
  };
  properties: Record<string, ModelPropertyDefinition>;
}

type ModelPropertyDefinition =
  | ModelPropertyStringDefinition
  | ModelPropertyNumberDefinition
  | ModelPropertyBooleanDefinition
  | ModelPropertyDateDefinition
  | ModelPropertyOneToOneDefinition
  | ModelPropertyManyToOneDefinition
  | ModelPropertyOneToManyDefinition
  | ModelPropertyManyToManyDefinition;

type ModelPropertyRelationalDefinition =
  | ModelPropertyOneToOneDefinition
  | ModelPropertyManyToOneDefinition
  | ModelPropertyOneToManyDefinition
  | ModelPropertyManyToManyDefinition;

interface CommonModelPropertyDefinition {
  required?: boolean;
}

interface ModelPropertyStringDefinition extends CommonModelPropertyDefinition {
  type: "string";
  default?: string;
}

interface ModelPropertyNumberDefinition extends CommonModelPropertyDefinition {
  type: "number";
  default?: number;
}

interface ModelPropertyBooleanDefinition extends CommonModelPropertyDefinition {
  type: "boolean" | "bool";
  default?: boolean;
}

interface ModelPropertyDateDefinition extends CommonModelPropertyDefinition {
  type: "date";
  default?: Date;
}

type ModelPropertyOneToOneDefinitionOptions =
  | ModelPropertyOneToOneDefinitionOptionsA
  | ModelPropertyOneToOneDefinitionOptionsB;

interface ModelPropertyOneToOneDefinitionOptionsA {
  /* The column on the foreign model that references the local model */
  foreignKey: string;
  /* When the foreign field references the local model, this is the field the local references it by. Defaults to "id" */
  key?: string;
}

interface ModelPropertyOneToOneDefinitionOptionsB {
  /* The column on the local model that references the foreign model */
  key: string;
  keyType?: string;
  /* When a local field references a foreign model, this is the field the foreign model is referenced by. Defaults to "id" */
  foreignKey?: string;
}

interface ModelPropertyOneToOneDefinition
  extends CommonModelPropertyDefinition {
  type: "one-to-one";
  modelId: string;
  default?: ModelDefinition;
  oneToOneOptions: ModelPropertyOneToOneDefinitionOptions;
}

interface ModelPropertyManyToOneDefinition
  extends CommonModelPropertyDefinition {
  type: "many-to-one";
  modelId: string;
  manyToOneOptions: {
    /* The field on local models that reference a foreign model */
    key: string;
    keyType?: string;
    /* When we need to reference a foreign model, we access the foreign field via this field name. Defaults to "id" */
    foreignKey?: string;
  };
}

interface ModelPropertyOneToManyDefinition
  extends CommonModelPropertyDefinition {
  type: "one-to-many";
  modelId: string;
  oneToManyOptions: {
    /* The field on the foreign models that reference a local model */
    foreignKey: string;
    /* When a foreign model in this relationship references the local model, it does so via this field name. Defaults to "id" */
    key?: string;
  };
}

interface ModelPropertyManyToManyDefinition
  extends CommonModelPropertyDefinition {
  type: "many-to-many";
  modelId: string;
  manyToManyOptions: {
    /* When looking up foreign models, foreignKey is the column we perform lookups by */
    foreignKey?: string;
    /* When looking up local models, we look them up via this column */
    localKey?: string;
    /* Many-to-many relationships require a join table */
    joinTable: {
      /* Canonical name of the join table */
      name: string;
      /* When referencing local models, this is the column we look things up by. Default: "local_model_id" */
      localModelKey?: string;
      /* When references foreign models, this is the column we look things up by. Default: "foreign_model_id" */
      foreignModelKey?: string;
    };
  };
}

type ModelId = string;

function isRelationshipType(
  type: string
): type is "one-to-one" | "many-to-one" | "one-to-many" | "many-to-many" {
  if (type === "one-to-one") return true;
  if (type === "many-to-one") return true;
  if (type === "one-to-many") return true;
  if (type === "many-to-many") return true;
  return false;
}

function assertIsModelSchemaType(
  modelSchemaDefinition: ModelPropertyDefinition
): asserts modelSchemaDefinition is ModelPropertyRelationalDefinition {
  if (!isRelationshipType(modelSchemaDefinition.type)) {
    throw new Error(
      `Unexpected invariance violation: assertIsModelSchemaType expected a relational model type. Got instead ${modelSchemaDefinition.type}`
    );
  }
}

export async function generateModels(schema: Schema): Promise<string> {
  let output: string = `const schema = ${JSON.stringify(schema)};\n\n`;

  const models: Map<ModelId, SchemaModel> = new Map();
  const relationshipTokens: Map<ModelId, [string, string][]> = new Map();

  for (const model of schema.models) {
    models.set(model.id, model);
    for (const modelProperty of Object.keys(model.properties)) {
      if (isRelationshipType(model.properties[modelProperty].type)) {
        const modelDefinition = model.properties[modelProperty];
        assertIsModelSchemaType(modelDefinition);
        if (!relationshipTokens.has(model.id)) {
          relationshipTokens.set(model.id, []);
        }
        relationshipTokens
          .get(model.id)
          .push([modelProperty, modelDefinition.modelId]);
      }
    }
  }

  const relationships: Map<SchemaModel, Map<string, SchemaModel>> = new Map();
  for (const modelId of relationshipTokens.keys()) {
    if (!models.has(modelId)) {
      continue;
    }
    const localModel = models.get(modelId)[0];
    if (!relationships.has(localModel)) {
      relationships.set(localModel, new Map());
    }
    for (const [modelProperty, foreignId] of relationshipTokens.get(modelId)) {
      if (!models.has(foreignId)) {
        throw new ModelRelationshipNotFoundError({
          models,
          onModel: modelId,
          modelProperty,
          notFoundId: foreignId,
        });
      }
      relationships
        .get(localModel)
        .set(modelProperty, models.get(foreignId)[0]);
    }
  }

  for (const [, modelOptions] of models.entries()) {
    // Without hidden/relational properties
    output += `interface ${modelOptions.names.singularPascal} {`;
    for (const accessor of Object.keys(modelOptions.properties)) {
      const property = modelOptions.properties[accessor];
      if (isRelationshipType(property.type)) {
        continue;
      }
      output += `  ${accessor}: ${property.type};`;
    }
    output += "}\n\n";

    // Raw DB object
    output += `interface Raw${modelOptions.names.singularPascal} {`;
    for (const accessor of Object.keys(modelOptions.properties)) {
      const property = modelOptions.properties[accessor];
      if (isRelationshipType(property.type)) {
        if (property.type === "many-to-one") {
          if (property.manyToOneOptions.key) {
            output += `  ${property.manyToOneOptions.key}: ${
              property.manyToOneOptions.keyType || "string"
            };`;
          }
        }
        if (property.type === "one-to-one") {
          if (property.oneToOneOptions.key) {
            output += `  ${property.oneToOneOptions.key}: ${
              (property.oneToOneOptions as Record<string, any>).keyType ||
              "string"
            };`;
          }
        }
        continue;
      }
      output += `  ${accessor}: ${property.type};`;
    }
    output += "}\n\n";
  }

  return output;
}

class ModelRelationshipNotFoundError extends Error {
  constructor({
    models,
    onModel,
    notFoundId,
    modelProperty,
  }: {
    models: Map<ModelId, SchemaModel>;
    onModel: string;
    notFoundId: string;
    modelProperty: string;
  }) {
    const sourceModel = models.get(onModel);
    const modelName = sourceModel.names.singularPascal;
    super(
      `Error parsing the ${modelName} model! Relationship ${modelName}.${modelProperty} references a model whose ID we have no record of (${notFoundId}). Check the spelling of the ID and make sure that the relationship model exists in the schema. All models: ${Array.from(
        models.keys()
      ).join(", ")}`
    );
  }
}
