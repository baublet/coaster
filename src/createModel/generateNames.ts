export function generateNames(
  name: string
): {
  singularCamel: string;
  singularPascal: string;
  pluralCamel: string;
  pluralPascal: string;
} {
  return {
    singularCamel: name,
    pluralCamel: name,
    pluralPascal: name,
    singularPascal: name,
  };
}
