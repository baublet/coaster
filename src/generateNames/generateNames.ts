export function generateNames(
  name: string
): {
  raw: string;
  rawCamel: string;
  rawPascal: string;
  singularCamel: string;
  singularPascal: string;
  pluralCamel: string;
  pluralPascal: string;
} {
  return {
    raw: name,
    rawCamel: name,
    rawPascal: name,
    singularCamel: name,
    pluralCamel: name,
    pluralPascal: name,
    singularPascal: name,
  };
}
