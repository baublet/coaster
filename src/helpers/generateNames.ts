import { camelCase, pascalCase, snakeCase } from "change-case";
import pluralize from "pluralize";
import memoize from "lodash.memoize";

export interface GeneratedNames {
  /**
   * The canonical name. Not safe for use in databases or in JS.
   */
  canonical: string;
  /**
   * Plural version of canonical.
   */
  plural: string;
  /**
   * Often the same as above, but in cases where canonical needs to vary from
   * the canonical, use this.
   */
  original: string;
  /**
   * Pluralized version of original.
   */
  originalPlural: string;
  /**
   * A database- and javascript-safe version of the string. Suitable for using
   * as database column names.
   */
  safe: string;
  /**
   * The plural version of the safe name.
   */
  pluralSafe: string;
  /**
   * The pascal-case name
   */
  pascal: string;
  /**
   * The camel-case name
   */
  camel: string;
}

/**
 * Takes a string and returns a structure of various names for use within an
 * application. If `original` is passed in, we use that as the original value.
 * This is useful if the name passed in is already transformed in some way.
 * @param name
 * @param original
 */
function generateNames(
  name: string,
  original: string | false = false
): GeneratedNames {
  const safe = snakeCase(name);
  return {
    canonical: name,
    original: original || name,
    originalPlural: pluralize(original || name, 2),
    plural: pluralize(name, 2),
    pluralSafe: pluralize(safe, 2),
    safe: safe,
    pascal: pascalCase(name),
    camel: camelCase(name)
  };
}

export default memoize<typeof generateNames>(generateNames);
