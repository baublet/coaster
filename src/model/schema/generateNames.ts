import toSnakeCase from "snake-case";

export interface SchemaNodeNames {
  canonical: string;
  safe: string;
}

export default function generateNames(name: string): SchemaNodeNames {
  return {
    canonical: name,
    safe: toSnakeCase(name)
  };
}
