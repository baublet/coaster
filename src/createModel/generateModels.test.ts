import { string } from "yargs";
import { generateModels } from "./generateModels";
import { generateNames } from "./generateNames";

it("generates the right types", async () => {
  await expect(
    generateModels({
      name: "Test Schema 1",
      models: [
        {
          id: "user",
          names: generateNames("User"),
          properties: {
            id: {
              type: "string",
            },
            bio: {
              type: "string",
            },
            created: {
              type: "date",
            },
          },
        },
        {
          id: "org",
          names: generateNames("Organization"),
          properties: {
            id: {
              type: "string",
            },
            description: {
              type: "string",
            },
            users: {
              type: "many-to-many",
              modelId: "user",
              manyToManyOptions: {
                joinTable: {
                  name: "user_orgs",
                },
              },
            },
          },
        },
        {
          id: "settings",
          names: generateNames("Settings"),
          properties: {
            id: {
              type: "string",
            },
            user: {
              type: "one-to-one",
              modelId: "user",
              oneToOneOptions: {
                key: "user_id",
              },
            },
          },
        },
      ],
    })
  ).resolves.toMatchSnapshot();
});
