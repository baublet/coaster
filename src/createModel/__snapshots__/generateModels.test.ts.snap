// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates the right types 1`] = `
"const schema = {\\"name\\":\\"Test Schema 1\\",\\"models\\":[{\\"id\\":\\"user\\",\\"names\\":{\\"singularCamel\\":\\"User\\",\\"pluralCamel\\":\\"User\\",\\"pluralPascal\\":\\"User\\",\\"singularPascal\\":\\"User\\"},\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\"},\\"bio\\":{\\"type\\":\\"string\\"},\\"created\\":{\\"type\\":\\"date\\"}}},{\\"id\\":\\"org\\",\\"names\\":{\\"singularCamel\\":\\"Organization\\",\\"pluralCamel\\":\\"Organization\\",\\"pluralPascal\\":\\"Organization\\",\\"singularPascal\\":\\"Organization\\"},\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\"},\\"description\\":{\\"type\\":\\"string\\"},\\"users\\":{\\"type\\":\\"many-to-many\\",\\"modelId\\":\\"user\\",\\"manyToManyOptions\\":{\\"joinTable\\":{\\"name\\":\\"user_orgs\\"}}}}},{\\"id\\":\\"settings\\",\\"names\\":{\\"singularCamel\\":\\"Settings\\",\\"pluralCamel\\":\\"Settings\\",\\"pluralPascal\\":\\"Settings\\",\\"singularPascal\\":\\"Settings\\"},\\"properties\\":{\\"id\\":{\\"type\\":\\"string\\"},\\"user\\":{\\"type\\":\\"one-to-one\\",\\"modelId\\":\\"user\\",\\"oneToOneOptions\\":{\\"key\\":\\"user_id\\"}}}}]};

interface User {  id: string;  bio: string;  created: date;}

interface RawUser {  id: string;  bio: string;  created: date;}

interface Organization {  id: string;  description: string;}

interface RawOrganization {  id: string;  description: string;}

interface Settings {  id: string;}

interface RawSettings {  id: string;  user_id: string;}

"
`;
